### DISCUSSION TITLE
Log file modification request

### DISCUSSION THREAD

**Discussion by rdica** [jamulussoftware/jamulus#1969] [Ideas]
Currently the log file format is CSV with fields: timestamp, IP, status
```
2021-08-22 22:39:43, xxx.xxx.xxx.xxx, connected (3)
2021-08-22 22:51:38,, server idling -------------------------------------
```

Is it possible to add field with nickname used on join?

**Comment by pcar75** [comment]
I strongly support this request !
Thanks.

**Comment by gilgongo** [comment]
I think this has been requested before. At the time, the mood was that Jamulus wants to be as anonymous as possible (as per https://jamulus.io/wiki/Privacy-Statement). 

But things change, so can you elaborate on why you would want to have names recorded in the logs? It might make the debate a bit easier.

**Reply by ann0see** [reply]
... and I still support being as anonymous as possible (privacy by design). If this is implemented it needs to be off by default.

**Reply by rdica** [reply]
As a server operator my goal is to provide players with an environment conducive to interruption- or distraction-free playing. I have been running my own server for 3-4 months now and have seen some of the negative side of behaviour.  I now watch the server log while playing in order to try and catch the IP of offensive users.
I manually add that IP to the servers firewall to block any further interaction with jamulus.

Using tools such as `whois` and `host/dig`, I try to confirm the owner (ISP, not actual user) of an IP compared to the country flag (if any) the user has set in their profile. Its a very imperfect process, made even more difficult if the user didn't set a location in their profile, thus making confirmation of an offending IP nearly impossible.

I don't always have the opportunity to notice the IP of such a user at the time they join and start causing problems, but I do take note of the username they used. At present there's no way to tie that username to an IP for later processing, like adding the IP to a firewall, hence my request.

To be clear, I'm not requesting keeping track of username changes in the server log, just the username they use when they first join.

I do understand the current position in keeping things as anonymous as possible, I feel however that server operators should be given a little more information to better manage the service they provide.

Thanks

**Reply by gilgongo** [reply]
Thanks for the explanation. The issue of user removal has also come up from time to time (sometimes related to the wider issue of user authentication). 

I assume you are talking about a public server. When user admin is mentioned, the responses generally are:

1. Users can simply solo the people they want to play with (or mute an offender). This is the preferred method becuase it is in keeping with the spirit of Jamulus in public mode: everyone should be equal and in control. After all, one person's disruption might be another's amusing joke, or perhaps unintended problems with equipment. Humans being human.

2. Use a private server instead. And one technique for dealing with persistent disruptors being to simply restart a public server in private mode once the desired members have joined. Existings connections will still work until the users explicitly diconnect their clients. 

4. Write an authentication layer into Jamulus. There have been various proposals to do this but I'm not in a position to judge them. I think there may even be third party systems/proofs of concept you could look at. The field of education is of course interested in this.

Of course, adding a user name to the logs may not be a technically difficult thing to do. Personally, I wouldn't oppose it, but agree with @ann0see that a) it should be off by default and b) if turned on, those who join the server should be told that their names are being logged for purposes determined by the server admin, given an opporunity not to join for that reason.

BTW  see also https://github.com/jamulussoftware/jamulus/discussions/1710#discussioncomment-910536 and https://github.com/jamulussoftware/jamulus/discussions/1371

**Reply by rdica** [reply]
Thanks for the response. I'm aware of the comments etc and potential solutions for user management, but they fall outside the scope of my request. 

@ann0see I could see a) being turned off by default, but I don't understand the thinking behind b).

Users aren't notified their IP addresses are being logged, so why should they be notified that a username would be as well?
The names aren't static, don't require registration, they're just arbitrary strings. IP addresses can reveal far more information about a user than any username scheme can.

**Reply by ann0see** [reply]
Yes. And that’s another "feature" I dislike. The server should not log anything potentially personal at all. I might be a bit strict on privacy ;-).

Maybe the JSON API PR https://github.com/jamulussoftware/jamulus/pull/1975 could be expanded in future.

I know that it IS possible to get the names with https://explorer.jamulus.io and that this might even be too public for some users.

**Reply by gilgongo** [reply]
>  I don't understand the thinking behind b).

Simply that unlike an IP address, you could inadvertenly reveal your identity by either giving your name or some other identifer (eg in the course of a text chat) that could then be recorded without your knowledge or consent. It's worth considering too that in the EU and other places (including the UK and US), if a server operator is able to link IP addresses to personally identiiable data, they then have obgliations under law that they would not have otherwise. So it might get complicated regarding things like data protection, subject access requests and so on. Sure, all very unlikely, but possible.

**Reply by rdica** [reply]
As things now stand, a server operator can associate a user with an IP. How does that bypass EU GDPR compliance, or non-EU data privacy compliance?

**Reply by gilgongo** [reply]
@rdica Sorry, not sure I undertand your question. Associating a user with an IP doesn't bypass GDPR compliance.

Fundamentally though, if you want to run a public server and don't want to be bothered by people who bother you,  the current prescription is that people can just solo those they want to hear, or mute them. The [Jamulus design principle](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#keep-it-simple-and-stupid-and-do-one-thing-and-do-it-well) here is that if something can be achieved already by other means, the bar for a new feature to do it is correspondingly higher.

**Reply by rdica** [reply]
I wasn't asking to have a feature added, or for user management to be implemented, just a modification to a log file that already exists. This has no bearing on client behaviour/use (muting or soloing), but can potentially facillitate third party user management by a server operator.
In your own words you suggest it's not difficult technically to implement, but there's some concern regarding logging usernames. Since logging is already occurring, I'm not understanding the additional concern.

I don't have anything more to add, per your request I have outlined why I would desire the change, thank you for your consideration. :)

**Reply by pcar75** [reply]
For the sake of completeness, if not coherency, let's consider the following ! 
In server mode (not headless) the 'server console' displays the complete IP address, the nickname and the number of server-side jitter buffers of each participants.  When recording, each participant's filename contains his/her nickname, IP address (first 3 numbers), port number, connection number and a mono/stereo value.  So I can't see the logic in penalizing a 'headless' Jamulus server by omitting the nickname from logs ? 
The current request is certainly not a breach in security or anonymity, could help further in setups and admin.  Obviously it would entail very little change ; it's has probably been encoded for a while in test/debug versions.
Thanks, hopefully .

**Reply by gilgongo** [reply]
To which it's worth adding that in non-headless server mode the nickname is **not** recorded in a log for later analysis, but when nickname **is** stored (in recording mode) users are alerted to the fact that recording is on.

The current request could therefore be seen as "a breach in anonymity" because we actually make a point of saying "The server does not otherwise store or record your Profile information" in [the privacy statement](https://jamulus.io/wiki/Privacy-Statement). I shall leave the legal point about IP/name matching capability alone, suffice to say it's a consideration.

I don't dispute that adding the nickname to the log might be trivial. My point is that Jamulus is, as @ann0see says, private by design. So if nickname storage were to be introduced, users would need to know before they joined the server. 

Be that as it may, the ask here it to make it easier to kick out troublesome users. But Jamulus provides a method for ignoring such users anyway so (and by the general [design principles](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#jamulus-projectsource-code-general-principles)) I don't think tampering with privacy for that is warranted.

And for the sake of completeness: work is being done on an authentication layer of some kind for Jamulus, so user idenfication in principle isn't completely off the cards in the long term.

**Reply by rdica** [reply]
US$0.02...

```
Audio Recordings
You will see a notice if you are connected to a Jamulus server while server recording is turned on. Recordings of each player are stored by the server separately as .WAV files and only the server operator has access to them.
```

Whats not mentioned in the privacy statement is the stored filename contains the username, src port, and first three octets of an IPv4 IP (the last octet easily found in the log file) as well. Being informed that a recording is taking place doesn't inform a user of the privacy implications. A typical user has no idea whats happening on the server during a recording.

There's a mention of `only the server operator has access to them`, is that meant to imply some added security level not previously defined? Consequently only a server operator has access to the log file as well, so why can't it contain a username as well? (There's actually two assumptions here; that the recordings, and logs are only accessible by a server operator, but this would be a discussion for another day.)

I accept that the team will not implement my request. You may want to update your privacy statement to better inform users what exactly is being stored on disk during a recording.

Thanks again for considering my request :)

**Reply by gilgongo** [reply]
> You may want to update your privacy statement to better inform users what exactly is being stored on disk during a recording.

Good point.

**Comment by Rob-NY** [comment]
I understand all sides on this one.  I do tend to lean more toward the server-operator side and the need for the right information in order to quickly and accurately identify clients for whatever purpose (bad actors, etc).    There are several references to public servers in the threads above, but there's really no such thing -- only listed and unlisted.  And since Jamulus is the underlying system in both cases, any reference to "privacy statements" for the underlying software (as opposed to the implementation of the software) is probably not the best idea anyway.  Additionally, the privacy policy as it is written today gives third-parties more information disclosure rights than it gives the server operators.  Frankly, the ability to do "connection-less" queries of a server exposes more information to more people than this logging request ever would.

I completely support the prior changes (like removing the IP from client responses) to protect anonymity at the client level.  However, I do think that fair consideration needs to be given to requests like this that are only relevant to the sysop for management and administrative purposes.

Further, plenty of systems allow for configuration options that might require storage disclosure to end users (Apache, etc).  I think what this entire thread really exposes is a potential issue in the disclosure at the CENTRAL SERVER or the PROFILE SETUP level -- which is a much bigger topic that this log file question.

Perhaps a realistic option is to add a general disclosure on the PROFILE screen to the effect that:  "Servers may log [your IP address and] your supplied name for server management and administration purposes.", and to modify the privacy statement for the software to include similar language.   The logging of session information is not uncommon or unreasonable.

**Reply by gilgongo** [reply]
@Rob-NY Thanks for this perspective. You say, "accurately identify clients for whatever purpose (bad actors, etc)". As I've said, the "bad actors" case is already handled by being able to mute or solo. The Jamulus design principle is, "If a feature or function can be achieved in another way by another system or method, then that is preferable to building that feature into Jamulus." 

So what other use cases are we suggesting here?

**Comment by Rob-NY** [comment]
@gilgongo - I appreciate your consideration on this.  There are several other use cases.   Offensive names and offensive chats are two that are not solved with the mute/solo option.  Then, there is the economic issue.

I believe Jamulus has evolved and is being used in ways that the original framers did not or could not imagine.  And to be clear, I completely support the free environment (free as in freedom, not free as in beer) that the directory infrastructure provides to the community.  However, at the end of the day someone is paying for these servers.   Bad actors consume a seat, bandwidth, and disk space -- and there is an economic reality to that.   A recorded Jamulus session with 10 songs and 10 players could easily be 5 gig.  If only one of those channels is a bad-actor, then ~500mb is attributed to them --  whether mute/solo was activated or not.   So now the system operators are paying to store and ultimately transmit this data to the ultimate 'consumer' of the recording.   At scale, these could turn into pretty big numbers.

Again, I want to be clear that your point is completely valid and I (sort of) agree with you with respect to servers that are listed on the directory servers as they exist today.   Unfortunately there isn't much choice right now if you want your server 'listed' as part of core.   And as I said before, this isn't really about the log file -- it is about operational control of a server that is being used in ways that were not necessarily envisioned in the earlier days of Jamulus.

One option might be to extend the directory server structure to include one or more "Managed Server" directory servers that are genre agnostic.  This would provide fair warning to users about the intended use of the servers (not necessarily private, but also not the Caligula-like environment that is desired on the public side today).  At this point you could then lock down the logging option to be 'prohibited' if registering with a directory server in class "A" (the current list), but permissible if connecting to any other directory server or as an unlisted server.  This would provide a reasonable balance between anonymity and management need -- and would greatly improve the ability to have a "privacy statement" that is scope appropriate (ie, Dir Servers that are part of the Jamulus infrastructure) and enforceable (with the code option restrictions).

I think this is an important discussion because it touches on several themes that have come up before with respect to the original Jamulus vision and the reality of present day usage.  I really appreciate the engagement here.

**Reply by gilgongo** [reply]
So by providing the user name in the logs, the admin could look at that log to determine which IP to block (assuming they had external tools to do that on the server) in the event of abuse not limited to just audio. 

On the wider issue of "operational control" of servers: I've mentioned there are currently efforts to build methods of managing access as long-term projects. Ideas like "class A" (or "moderated"?) directories could be part of that perhaps. Maybe if you started a separate discussion of that idea?

I would imagine though that if somebody wants to do a PR for showing the name in the log, then that will happen eventually. It is usually true that control wants more control.

**Reply by Rob-NY** [reply]
PR #2010 has been created for discussion purposes and to provide context for the asks within this thread.

**Comment by ghost** [comment]
New Idea:
- forget the log, and go for what you actually need.
- make a gui window for a convenient nonpermanent white list.
- gui shows ipaddrs and pseudonyms to whitelist in the moment,
others disconnected immediately on menu exit. (or never connected)
- zero user-friendiness (whitelist is activated once and disappears)
- (session restart if you mess up. This is not an App it's Linux)
- pencil and paper required to log anything.
(make it inconvenient for screenshots etc.)

**Reply by Rob-NY** [reply]
I'm not sure I understand what you're saying, exactly.  If you're suggesting some type of mechanism where you can manage allowed or disallowed IP addresses, then I think everyone is in agreement.  This thread seems to be centered around getting the information that you list in your third bullet. Personally, this is exactly what I've done with my vJammers.com solution.  Below are two screen shots - one with the server in 'locked' mode, and the other in 'open' mode.  Both let you toggle folks in or out.  

The 'blocking' portion of this application works at an OS firewall level and does not require any changes to Jamulus, per se, but it does require PR #2006 in order to get the complete connection information.   The complete connection information is your third bullet point, and getting this information in the log is the root of this entire thread.

![Screenshot from 2021-09-09 21-40-57](https://user-images.githubusercontent.com/72955397/132785271-c17c6523-74df-4ce9-a014-25f3ce1c84ab.png)

![Screenshot from 2021-09-09 21-43-47](https://user-images.githubusercontent.com/72955397/132785668-84e6d24a-3933-4aa5-a126-1c8a4ec62895.png)

**Reply by ghost** [reply]
My points above were made quickly as an idea came to me. The points are intended to adhere to the KISS principle. I also mentioned a whitellist (not a black list because of privacy concerns etc.). The idea is similar to a stealth firewall. Furthermore, people that want 'metrics' and 'digging down to root causes' and 'data mining' and other stupid 'management stuff' should sit in the corner wearing their suit and tie.
Also bubble head, gum chewing App swipers, and 'into the fountain' texters need to understand that they are dumb-users with their smart-apps (this is a passing fad so it cannot be incorporated anywhere).

BTW what you have done looks real good (maybe too deluxe - less is more).

**Reply by ghost** [reply]
The log exposes sensitive data between Jamulus and vJammers.
Possibly, one could use a pencil and paper to write down the relevant information
(by talking to their friends or even seeing what little one-line info Jamulus may present).
This hand written information can be compiled [1] into a command line option (or signal)
so that it can be passed to Jamulus and/or vJammers.
1- hand compiled

**Reply by Rob-NY** [reply]
You definitely sparked a good laugh...   I took off my tie...

**Reply by ghost** [reply]
Thank you. I'm happy and fair game for a good laugh.

**Comment by JaminShanti** [comment]
Saw you talking about this in the Jamulus channel.  I do think the a server admin ends more tools to track problematic clients.  The Public server is a valuable free service.  It's not like you can't find this info out based on netstat info and figure out what is who.  Obscurity is not good security.

**Reply by ghost** [reply]
This is about accountability. If Jamulus forces you to find the information yourself, you are accountable.
There is no need to track problematic clients when you are using a whitelist.
There may be a problem for large choirs. However, where there is a will, you will find a way.

**Reply by JaminShanti** [reply]
Whitelist?  With Musicians.  I don't really like whitelist unless your talking about the admin being able to generate it from connected users, then banning anyone not no the list.  You know you don't really need to ban them.  Just Mute them, they can listen all they want.  I see groups today doing this.  They solo everyone in their group so the rest of the room is ignored.

**Reply by ghost** [reply]
- The whitelist is indeed about the admin being able to generate it from connected users.
Note that no one is banned here (just that not everyone is accepted). That is what a whitelist does.
- A whitelist does not know anything But your friends. This way the people that are not on the whitelist cannot complain to the law that they couldn't crash your party. Now if You take pot-shots with your shotgun, or picture-shots with your camera, or ipaddr shots... you might get yourself in trouble.
- It would be foolish of Jamulus to take potshots or track the bad actors
- Jamulus must be legal, and must not misrepresent itself with incorrect terminology. Any false claims will be vigilantly exposed forthwith.

**Comment by JaminShanti** [comment]
In Software products you add features to attract new customers and add security options to retain customers.  Something to remember.
